crates/fuel-core/src/graphql_api/storage/balances.rs

```diff
@@ -0,0 +1,77 @@
+use fuel_core_storage::{
+    blueprint::plain::Plain,
+    codec::{
+        postcard::Postcard,
+        raw::Raw,
+    },
+    structured_storage::TableWithBlueprint,
+    Mappable,
+};
+use fuel_core_types::{
+    fuel_tx::{
+        Address,
+        AssetId,
+    },
+    fuel_vm::double_key,
+};
+use rand::{
+    distributions::Standard,
+    prelude::Distribution,
+    Rng,
+};
+
+pub type ItemAmount = u64;
+pub type TotalBalanceAmount = u128;
+
+double_key!(BalancesKey, Address, address, AssetId, asset_id);
+impl Distribution<BalancesKey> for Standard {
+    fn sample<R: Rng + ?Sized>(&self, rng: &mut R) -> BalancesKey {
+        let mut bytes = [0u8; BalancesKey::LEN];
+        rng.fill_bytes(bytes.as_mut());
+        BalancesKey::from_array(bytes)
+    }
+}
+
+impl core::fmt::Display for BalancesKey {
+    fn fmt(&self, f: &mut core::fmt::Formatter) -> core::fmt::Result {
+        write!(f, "address={} asset_id={}", self.address(), self.asset_id())
+    }
+}
+
+/// This table stores the balances of coins per owner and asset id.
+pub struct CoinBalances;
+
+impl Mappable for CoinBalances {
+    type Key = BalancesKey;
+    type OwnedKey = Self::Key;
+    type Value = TotalBalanceAmount;
+    type OwnedValue = Self::Value;
+}
+
+impl TableWithBlueprint for CoinBalances {
+    type Blueprint = Plain<Raw, Postcard>;
+    type Column = super::Column;
+
+    fn column() -> Self::Column {
+        Self::Column::CoinBalances
+    }
+}
+
+/// This table stores the balances of messages per owner.
+pub struct MessageBalances;
+
+impl Mappable for MessageBalances {
+    type Key = Address;
+    type OwnedKey = Self::Key;
+    type Value = TotalBalanceAmount;
```

Comment:
We have basically 2 balances for messages: retryable balance and non retryable balance.

The message has retryable amount if its data is not empty, and a non-returnable amount otherwise.

So the type here should be `(u128, u128)` or a structure with corresponding fields.

You can see how `AssetsQuery` works and that it skips messages if data field is not empty to show non retryable balance.

Reply:
The following changes have been introduced:

* Additional test coverage was added in 1d4005673386e8dbb92ec48138c2dc338bf1f291 (these tests are currently failing with the new implementation, but pass with the classic ones)
* The `balances()` function in off_chain has been refactored in bf8334637c2dc95f2f019e4c814f1d0fd8f37026

Next steps is to split message balance into two parts to track non-, vs retryable balance. This will also break the generalization of balance update code, so I'll make it less-generic but more simple.